eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala

```diff
@@ -839,19 +841,28 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder, remo
     case Event(revocation: RevokeAndAck, d: DATA_NORMAL) =>
       // we received a revocation because we sent a signature
       // => all our changes have been acked
-      Commitments.receiveRevocation(d.commitments, revocation) match {
-        case Right((commitments1, forwards)) =>
+      Commitments.receiveRevocation(d.commitments, revocation, nodeParams.onChainFeeConf.feerateToleranceFor(remoteNodeId).maxDustHtlcExposure) match {
+        case Right((commitments1, actions)) =>
           cancelTimer(RevocationTimeout.toString)
           log.debug("received a new rev, spec:\n{}", Commitments.specs2String(commitments1))
-          forwards.foreach {
-            case Right(forwardAdd) =>
-              log.debug("forwarding {} to relayer", forwardAdd)
-              relayer ! forwardAdd
-            case Left(result) =>
+          actions.foreach {
+            case PostRevocationAction.RelayHtlc(add) =>
+              log.debug("forwarding incoming htlc {} to relayer", add)
+              relayer ! Relayer.RelayForward(add)
+            case PostRevocationAction.RejectHtlc(add) =>
+              log.debug("rejecting incoming htlc {}", add)
+              // NB: we don't set commit = true, we will sign all updates at once afterwards.
+              self ! CMD_FAIL_HTLC(add.id, Right(TemporaryChannelFailure(d.channelUpdate)))
+            case PostRevocationAction.RelayFailure(result) =>
               log.debug("forwarding {} to relayer", result)
               relayer ! result
           }
-          if (Commitments.localHasChanges(commitments1) && d.commitments.remoteNextCommitInfo.left.map(_.reSignAsap) == Left(true)) {
+          val signAsap = actions.exists {
+            case _: PostRevocationAction.RejectHtlc => true
+            case _: PostRevocationAction.RelayHtlc => false
+            case _: PostRevocationAction.RelayFailure => false
+          } || (Commitments.localHasChanges(commitments1) && d.commitments.remoteNextCommitInfo.left.map(_.reSignAsap) == Left(true))
+          if (signAsap) {
```

Comment:
I would remove this and do `CMD_FAIL_HTLC(..., commit = true)` above. It's a different concern.

Reply:
It's a different concern, but we want the same mechanism: it doesn't make sense to send N `CMD_SIGN()` when we only need one?

Reply:
But maybe it doesn't make sense to try to optimize this, the redundant `CMD_SIGN` will just be ignored...

Reply:
Done in https://github.com/ACINQ/eclair/pull/1985/commits/d0ffd73c1c1c49ceb1fd7c1b789fbfa9067da98f